File: VB/CustomRowInGrid/datasourceWrapper.vb
Imports System
Imports System.ComponentModel

Namespace GridCustomRows

    Friend Class DataSourceWrapper
        Implements IBindingList

        Private _neastedList As IBindingList

        Private _customRows As IBindingList

        Public Property NestedList As IBindingList
            Get
                Return _neastedList
            End Get

            Set(ByVal value As IBindingList)
                If _neastedList Is value Then Return
                _neastedList = value
                RaiseEvent ListChanged(Me, New ListChangedEventArgs(ListChangedType.Reset, 0))
            End Set
        End Property

        Public Property CustomRows As IBindingList
            Get
                Return _customRows
            End Get

            Set(ByVal value As IBindingList)
                If _customRows Is value Then Return
                _customRows = value
                RaiseEvent ListChanged(Me, New ListChangedEventArgs(ListChangedType.Reset, NestedList.Count))
            End Set
        End Property

        Public Sub New(ByVal target As IBindingList, ByVal custom As IBindingList)
            MyBase.New()
            _neastedList = target
            _customRows = custom
            AddHandler NestedList.ListChanged, New ListChangedEventHandler(AddressOf NeastedListChanged)
            AddHandler CustomRows.ListChanged, New ListChangedEventHandler(AddressOf NeastedListChanged)
        End Sub

        Private Sub NeastedListChanged(ByVal sender As Object, ByVal e As ListChangedEventArgs)
            If ListChangedEvent IsNot Nothing Then
                If sender Is NestedList Then
                    RaiseEvent ListChanged(Me, New ListChangedEventArgs(e.ListChangedType, e.NewIndex, e.OldIndex))
                Else
                    RaiseEvent ListChanged(Me, New ListChangedEventArgs(e.ListChangedType, e.NewIndex + NestedList.Count, e.OldIndex + NestedList.Count))
                End If
            End If
        End Sub

        Public Sub AddIndex(ByVal [property] As PropertyDescriptor) Implements IBindingList.AddIndex
            Throw New NotImplementedException()
        End Sub

        Public Function AddNew() As Object Implements IBindingList.AddNew
            Throw New NotImplementedException()
        End Function

        Public ReadOnly Property AllowEdit As Boolean Implements IBindingList.AllowEdit
            Get
                Return True
            End Get
        End Property

        Public ReadOnly Property AllowNew As Boolean Implements IBindingList.AllowNew
            Get
                Return True
            End Get
        End Property

        Public ReadOnly Property AllowRemove As Boolean Implements IBindingList.AllowRemove
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public Sub ApplySort(ByVal [property] As PropertyDescriptor, ByVal direction As ListSortDirection) Implements IBindingList.ApplySort
            Throw New NotImplementedException()
        End Sub

        Public Function Find(ByVal [property] As PropertyDescriptor, ByVal key As Object) As Integer Implements IBindingList.Find
            Throw New NotImplementedException()
        End Function

        Public ReadOnly Property IsSorted As Boolean Implements IBindingList.IsSorted
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public Event ListChanged As ListChangedEventHandler Implements IBindingList.ListChanged

        Public Sub RemoveIndex(ByVal [property] As PropertyDescriptor) Implements IBindingList.RemoveIndex
            Throw New NotImplementedException()
        End Sub

        Public Sub RemoveSort() Implements IBindingList.RemoveSort
            Throw New NotImplementedException()
        End Sub

        Public ReadOnly Property SortDirection As ListSortDirection Implements IBindingList.SortDirection
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public ReadOnly Property SortProperty As PropertyDescriptor Implements IBindingList.SortProperty
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public ReadOnly Property SupportsChangeNotification As Boolean Implements IBindingList.SupportsChangeNotification
            Get
                Return True
            End Get
        End Property

        Public ReadOnly Property SupportsSearching As Boolean Implements IBindingList.SupportsSearching
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public ReadOnly Property SupportsSorting As Boolean Implements IBindingList.SupportsSorting
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public Function Add(ByVal value As Object) As Integer Implements Collections.IList.Add
            Throw New NotImplementedException()
        End Function

        Public Sub Clear() Implements Collections.IList.Clear
            Throw New NotImplementedException()
        End Sub

        Public Function Contains(ByVal value As Object) As Boolean Implements Collections.IList.Contains
            Throw New NotImplementedException()
        End Function

        Public Function IndexOf(ByVal value As Object) As Integer Implements Collections.IList.IndexOf
            Dim index As Integer = CustomRows.IndexOf(value)
            If index = -1 Then Return NestedList.IndexOf(value)
            Return index + NestedList.Count
        End Function

        Public Sub Insert(ByVal index As Integer, ByVal value As Object) Implements Collections.IList.Insert
            Throw New NotImplementedException()
        End Sub

        Public ReadOnly Property IsFixedSize As Boolean Implements Collections.IList.IsFixedSize
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public ReadOnly Property IsReadOnly As Boolean Implements Collections.IList.IsReadOnly
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public Sub Remove(ByVal value As Object) Implements Collections.IList.Remove
            Throw New NotImplementedException()
        End Sub

        Public Sub RemoveAt(ByVal index As Integer) Implements Collections.IList.RemoveAt
            Throw New NotImplementedException()
        End Sub

        Default Public Property Item(ByVal index As Integer) As Object Implements Collections.IList.Item
            Get
                If index >= NestedList.Count Then
                    Return CustomRows(index - NestedList.Count)
                End If

                Return NestedList(index)
            End Get

            Set(ByVal value As Object)
                If index < CustomRows.Count Then
                    CustomRows(index) = value
                End If

                NestedList(index) = value
            End Set
        End Property

        Public Sub CopyTo(ByVal array As Array, ByVal index As Integer) Implements Collections.ICollection.CopyTo
            Throw New NotImplementedException()
        End Sub

        Public ReadOnly Property Count As Integer Implements Collections.ICollection.Count
            Get
                Return NestedList.Count + CustomRows.Count
            End Get
        End Property

        Public ReadOnly Property IsSynchronized As Boolean Implements Collections.ICollection.IsSynchronized
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public ReadOnly Property SyncRoot As Object Implements Collections.ICollection.SyncRoot
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public Function GetEnumerator() As Collections.IEnumerator Implements Collections.IEnumerable.GetEnumerator
            Throw New NotImplementedException()
        End Function

        Public Function IsCustomItem(ByVal value As Object) As Boolean
            Return CustomRows.Contains(value)
        End Function

        Public Function IsCustomItemIndex(ByVal index As Integer) As Boolean
            Return index >= NestedList.Count AndAlso index < CustomRows.Count + NestedList.Count
        End Function
    End Class
End Namespace

File: CS/CustomRowInGrid/CustomRowInGrid/datasourceWrapper.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.ComponentModel;

namespace GridCustomRows
{
    class DataSourceWrapper : IBindingList
    {
        IBindingList _neastedList;
        IBindingList _customRows;

        public IBindingList NestedList
        {
            get { return _neastedList; }
            set {
                if (_neastedList == value)
                    return;
                _neastedList = value;
                ListChanged(this, new ListChangedEventArgs(ListChangedType.Reset, 0));
            }
        }
        public IBindingList CustomRows
        {
            get { return _customRows; }
            set
            {
                if (_customRows == value)
                    return;
                _customRows = value;
                ListChanged(this, new ListChangedEventArgs(ListChangedType.Reset, NestedList.Count));
            }
        }

        public DataSourceWrapper(IBindingList target,IBindingList custom)
            : base()
        {
            _neastedList = target;
            _customRows = custom;
            NestedList.ListChanged+=new ListChangedEventHandler(NeastedListChanged);
            CustomRows.ListChanged += new ListChangedEventHandler(NeastedListChanged);
        }

        void NeastedListChanged(object sender, ListChangedEventArgs e)
        {
            if (ListChanged != null)
                if (sender == NestedList)
                    ListChanged(this, new ListChangedEventArgs(e.ListChangedType, e.NewIndex, e.OldIndex));        
                else
                    ListChanged(this, new ListChangedEventArgs(e.ListChangedType, e.NewIndex + NestedList.Count, e.OldIndex + NestedList.Count));
        }

        public void AddIndex(PropertyDescriptor property)
        {
            throw new NotImplementedException();
        }

        public object AddNew()
        {
            throw new NotImplementedException();
        }

        public bool AllowEdit
        {
            get { return true; }
        }

        public bool AllowNew
        {
            get { return true; }
        }

        public bool AllowRemove
        {
            get { throw new NotImplementedException(); }
        }

        public void ApplySort(PropertyDescriptor property, ListSortDirection direction)
        {
            throw new NotImplementedException();
        }

        public int Find(PropertyDescriptor property, object key)
        {
            throw new NotImplementedException();
        }

        public bool IsSorted
        {
            get { throw new NotImplementedException(); }
        }

        public event ListChangedEventHandler ListChanged;

        public void RemoveIndex(PropertyDescriptor property)
        {
            throw new NotImplementedException();
        }

        public void RemoveSort()
        {
            throw new NotImplementedException();
        }

        public ListSortDirection SortDirection
        {
            get { throw new NotImplementedException(); }
        }

        public PropertyDescriptor SortProperty
        {
            get { throw new NotImplementedException(); }
        }

        public bool SupportsChangeNotification
        {
            get { return true; }
        }

        public bool SupportsSearching
        {
            get { throw new NotImplementedException(); }
        }

        public bool SupportsSorting
        {
            get { throw new NotImplementedException(); }
        }

        public int Add(object value)
        {
            throw new NotImplementedException();
        }

        public void Clear()
        {
            throw new NotImplementedException();
        }

        public bool Contains(object value)
        {
            throw new NotImplementedException();
        }

        public int IndexOf(object value)
        {

            int index = CustomRows.IndexOf(value);
            if (index == -1)
                return NestedList.IndexOf(value);
            return index + NestedList.Count;
        }

        public void Insert(int index, object value)
        {
            throw new NotImplementedException();
        }

        public bool IsFixedSize
        {
            get { throw new NotImplementedException(); }
        }

        public bool IsReadOnly
        {
            get { throw new NotImplementedException(); }
        }

        public void Remove(object value)
        {
            throw new NotImplementedException();
        }

        public void RemoveAt(int index)
        {
            throw new NotImplementedException();
        }

        public object this[int index]
        {
            get
            {
                if (index >= NestedList.Count)
                {
                    return CustomRows[index - NestedList.Count];
                }
                return NestedList[index];
            }
            set
            {
                if (index < CustomRows.Count)
                {
                    CustomRows[index] = value;
                }
                NestedList[index] = value;
            }
        }

        public void CopyTo(Array array, int index)
        {
            throw new NotImplementedException();
        }

        public int Count
        {
            get { return NestedList.Count + CustomRows.Count; }
        }

        public bool IsSynchronized
        {
            get { throw new NotImplementedException(); }
        }

        public object SyncRoot
        {
            get { throw new NotImplementedException(); }
        }

        public System.Collections.IEnumerator GetEnumerator()
        {
            throw new NotImplementedException();
        }
        public bool IsCustomItem(object value)
        {
            return CustomRows.Contains(value);
        }
        public bool IsCustomItemIndex(int index)
        {

            return ((index >= NestedList.Count) && (index < CustomRows.Count + NestedList.Count));
        }
    }
}
